# Bank Loan Analysis

## Project Overview

The goal of this project is to analyze bank loan data to understand customer behavior, loan performance, and factors affecting loan status.

The project workflow includes:
- import libraries
- liading data
- Data Understanding
- Data Cleaning
- SQL Analysis
- Power BI Dashboard
- Business Insights

## import libraries

In [5]:
import numpy as np
import pandas as pd
import matplotlib as plt

## loading data

In [6]:
Bank = pd.read_csv(r"C:\My Project\Bank\financial_loan.csv")

## understandin data

In [7]:
Bank.head()

,id,address_state,application_type,emp_length,emp_title,grade,home_ownership,issue_date,last_credit_pull_date,last_payment_date,...,sub_grade,term,verification_status,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
0,1077430,GA,INDIVIDUAL,< 1 year,Ryder,C,RENT,11-02-2021,13-09-2021,13-04-2021,...,C4,60 months,Source Verified,30000.0,0.0100,59.83,0.1527,2500,4,1009
1,1072053,CA,INDIVIDUAL,9 years,MKC Accounting,E,RENT,01-01-2021,14-12-2021,15-01-2021,...,E1,36 months,Source Verified,48000.0,0.0535,109.43,0.1864,3000,4,3939
2,1069243,CA,INDIVIDUAL,4 years,Chemat Technology Inc,C,RENT,05-01-2021,12-12-2021,09-01-2021,...,C5,36 months,Not Verified,50000.0,0.2088,421.65,0.1596,12000,11,3522
3,1041756,TX,INDIVIDUAL,< 1 year,barnes distribution,B,MORTGAGE,25-02-2021,12-12-2021,12-03-2021,...,B2,60 months,Source Verified,42000.0,0.0540,97.06,0.1065,4500,9,4911
4,1068350,IL,INDIVIDUAL,10+ years,J&J Steel Inc,A,MORTGAGE,01-01-2021,14-12-2021,15-01-2021,...,A1,36 months,Verified,83000.0,0.0231,106.53,0.0603,3500,28,3835


In [8]:
Bank.columns

Index(['id', 'address_state', 'application_type', 'emp_length', 'emp_title',
       'grade', 'home_ownership', 'issue_date', 'last_credit_pull_date',
       'last_payment_date', 'loan_status', 'next_payment_date', 'member_id',
       'purpose', 'sub_grade', 'term', 'verification_status', 'annual_income',
       'dti', 'installment', 'int_rate', 'loan_amount', 'total_acc',
       'total_payment'],
      dtype='object')

In [9]:
Bank.shape

(38576, 24)

In [10]:
Bank.isnull().sum()

id                          0
address_state               0
application_type            0
emp_length                  0
emp_title                1438
grade                       0
home_ownership              0
issue_date                  0
last_credit_pull_date       0
last_payment_date           0
loan_status                 0
next_payment_date           0
member_id                   0
purpose                     0
sub_grade                   0
term                        0
verification_status         0
annual_income               0
dti                         0
installment                 0
int_rate                    0
loan_amount                 0
total_acc                   0
total_payment               0
dtype: int64

we have 1438 null values in emp_title column

In [11]:
Bank.duplicated().sum()

np.int64(0)

In [12]:
Bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38576 entries, 0 to 38575
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     38576 non-null  int64  
 1   address_state          38576 non-null  object 
 2   application_type       38576 non-null  object 
 3   emp_length             38576 non-null  object 
 4   emp_title              37138 non-null  object 
 5   grade                  38576 non-null  object 
 6   home_ownership         38576 non-null  object 
 7   issue_date             38576 non-null  object 
 8   last_credit_pull_date  38576 non-null  object 
 9   last_payment_date      38576 non-null  object 
 10  loan_status            38576 non-null  object 
 11  next_payment_date      38576 non-null  object 
 12  member_id              38576 non-null  int64  
 13  purpose                38576 non-null  object 
 14  sub_grade              38576 non-null  object 
 15  te

The date columns have the wrong data type.

In [13]:
Bank.describe()

,id,member_id,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
count,3.857600e+04,3.857600e+04,3.857600e+04,38576.000000,38576.000000,38576.000000,38576.000000,38576.000000,38576.000000
mean,6.810371e+05,8.476515e+05,6.964454e+04,0.133274,326.862965,0.120488,11296.066855,22.132544,12263.348533
std,2.113246e+05,2.668105e+05,6.429368e+04,0.066662,209.092000,0.037164,7460.746022,11.392282,9051.104777
min,5.473400e+04,7.069900e+04,4.000000e+03,0.000000,15.690000,0.054200,500.000000,2.000000,34.000000
25%,5.135170e+05,6.629788e+05,4.150000e+04,0.082100,168.450000,0.093200,5500.000000,14.000000,5633.000000
50%,6.627280e+05,8.473565e+05,6.000000e+04,0.134200,283.045000,0.118600,10000.000000,20.000000,10042.000000
75%,8.365060e+05,1.045652e+06,8.320050e+04,0.185900,434.442500,0.145900,15000.000000,29.000000,16658.000000
max,1.077501e+06,1.314167e+06,6.000000e+06,0.299900,1305.190000,0.245900,35000.000000,90.000000,58564.000000


Outlier Handling

After examining the numerical columns (`annual_income`, `loan_amount`, `interest_rate`, `installment`, `debt_to_income_ratio`, `total_payment`, and `total_accounts`) using descriptive statistics and boxplots, no outliers were removed.

This decision was made because the detected extreme values appear to represent **valid customer records rather than data entry errors**. In banking datasets, customers can naturally have very high incomes, large loan amounts, or high total payments. Removing these observations could eliminate important business cases and introduce bias into the analysis.

Only invalid or impossible values (e.g., negative income, negative loan amount, or unrealistic interest rates) would be considered for removal. Since no such values were identified, all records were retained for the analysis.

In [14]:
Bank["grade"].value_counts()

grade
B    11674
A     9689
C     7904
D     5182
E     2786
F     1028
G      313
Name: count, dtype: int64

In [15]:
Bank["loan_status"].value_counts()

loan_status
Fully Paid     32145
Charged Off     5333
Current         1098
Name: count, dtype: int64

In [16]:
Bank["home_ownership"].value_counts()

home_ownership
RENT        18439
MORTGAGE    17198
OWN          2838
OTHER          98
NONE            3
Name: count, dtype: int64

In [17]:
Bank[Bank["home_ownership"] == "NONE"]

,id,address_state,application_type,emp_length,emp_title,grade,home_ownership,issue_date,last_credit_pull_date,last_payment_date,...,sub_grade,term,verification_status,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
7744,231321,VA,INDIVIDUAL,5 years,ups,A,NONE,08-01-2021,11-01-2021,11-02-2021,...,A3,36 months,Not Verified,22200.0,0.0524,312.22,0.0775,10000,22,11240
31666,121535,MA,INDIVIDUAL,< 1 year,NaN,B,NONE,07-08-2021,07-08-2021,10-09-2021,...,B1,36 months,Not Verified,120000.0,0.1500,88.65,0.0870,2800,8,3191
31667,121568,CA,INDIVIDUAL,< 1 year,NaN,B,NONE,07-08-2021,07-08-2021,10-09-2021,...,B4,36 months,Not Verified,100000.0,0.1500,128.41,0.0964,4000,8,4622


In [18]:
Bank["home_ownership"]=Bank["home_ownership".replace("NONE","OTHER")]

In [19]:
Bank["home_ownership"].value_counts()

home_ownership
RENT        18439
MORTGAGE    17198
OWN          2838
OTHER          98
NONE            3
Name: count, dtype: int64

In [20]:
Bank["purpose"].value_counts()

purpose
Debt consolidation    18214
credit card            4998
other                  3824
home improvement       2876
major purchase         2110
small business         1776
car                    1497
wedding                 928
medical                 667
moving                  559
house                   366
vacation                352
educational             315
renewable_energy         94
Name: count, dtype: int64

Debt consolidation is the most common reason for loan applications, indicating that many customers use loans to combine existing debts into a single payment.


## cleaning data

In [21]:
Bank["emp_title"]=Bank['emp_title'].fillna("Unknown")

In [22]:
date_cols = [
    "issue_date",
    "last_credit_pull_date",
    "last_payment_date",
    "next_payment_date"
]

for col in date_cols:
    Bank[col] = pd.to_datetime(
        Bank[col],
        format="%d-%m-%Y",
        errors="coerce"
    )

In [23]:
Bank["emp_length"].value_counts()

emp_length
10+ years    8870
< 1 year     4575
2 years      4382
3 years      4088
4 years      3428
5 years      3273
1 year       3229
6 years      2228
7 years      1772
8 years      1476
9 years      1255
Name: count, dtype: int64

In [24]:
Bank["emp_length"] = (
    Bank["emp_length"]
    .replace("< 1 year", "1")
    .replace("10+ years", "10")
    .astype(int)
)

ValueError: invalid literal for int() with base 10: '9 years'

In [ ]:
Bank["emp_length"].value_counts()

emp_length
10    8870
1     7804
2     4382
3     4088
4     3428
5     3273
6     2228
7     1772
8     1476
9     1255
Name: count, dtype: int64

In [ ]:
Bank.rename(columns={"term": "loan_term_months"}, inplace=True)

In [ ]:
Bank.sample()

,id,address_state,application_type,emp_length,emp_title,grade,home_ownership,issue_date,last_credit_pull_date,last_payment_date,...,sub_grade,loan_term_months,verification_status,annual_income,dti,installment,int_rate,loan_amount,total_acc,total_payment
32234,503642,OH,INDIVIDUAL,10,mitchells salon and day spa,B,MORTGAGE,2021-04-10,2021-05-16,2021-05-13,...,B1,36 months,Not Verified,45000.0,0.2253,64.43,0.0988,2000,22,2319


In [ ]:
Bank["loan_term_months"].value_counts()

loan_term_months
36 months    28237
60 months    10339
Name: count, dtype: int64

In [ ]:
Bank["loan_term_months"]=Bank["loan_term_months"].str.replace("months","").astype(int)

In [ ]:
Bank["loan_term_months"].value_counts()

loan_term_months
36    28237
60    10339
Name: count, dtype: int64

In [ ]:
Bank.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38576 entries, 0 to 38575
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     38576 non-null  int64         
 1   address_state          38576 non-null  object        
 2   application_type       38576 non-null  object        
 3   emp_length             38576 non-null  int64         
 4   emp_title              38576 non-null  object        
 5   grade                  38576 non-null  object        
 6   home_ownership         38576 non-null  object        
 7   issue_date             38576 non-null  datetime64[ns]
 8   last_credit_pull_date  38576 non-null  datetime64[ns]
 9   last_payment_date      38576 non-null  datetime64[ns]
 10  loan_status            38576 non-null  object        
 11  next_payment_date      38576 non-null  datetime64[ns]
 12  member_id              38576 non-null  int64         
 13  p

In [ ]:
Bank_loan=Bank.to_csv("bank_loan_clean.csv", index=False)